Fix module merging crash due to wrongly computed extended nominal type.

varungandhi-apple opened the pull request:
Resolves [SR-11227](https://bugs.swift.org/browse/SR-11227) and rdar://problem/53712389.

Unlike  #26592, I have not added any desugaring here -- we do not increase the code we support. Instead we emit a warning asking the user to not write compositions in extended types.

~~First two commits are just minor cleanups. (Hmm, the commits are showing up in chronological order, not in topological order. The second commit in chronological order is the main one.)~~

~~There is a TODO left; I wasn't quite sure how to get the relevant source range for the extended type without computing the type itself.~~

varungandhi-apple committed:
[NFC] Rename 'baseFoo' to 'extendedTypeFoo' in (de)serialization.

varungandhi-apple committed:
[NFC] Add explicit #include for DiagnosticsSema.h in TypeCheckGeneric.cpp.

varungandhi-apple commented:
@swift-ci please smoke test

theblixguy commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1601,6 +1601,11 @@ ERROR(inferred_opaque_type,none,
       "non-nominal type %0 cannot be extended", (Type))
+WARNING(multiple_nominals_in_extended_type,none,
+        "expected a single nominal type in the extended type position but"
+        " found several. This may lead to unexpected errors while"
```
Nit: Could you move the space to the line above and remove the period at the end? (look at other multi-line diagnostics for example). Also, perhaps the wording of the warning could be simplified?

```
WARNING(multiple_nominals_in_extended_type,none,
        "extending several types at once via protocol composition " 
        "may lead to unexpected errors while type-checking. Use a "
        "'where' clause for additional constraints", ())

varungandhi-apple replied:
I fully agree with the sentiment that the current diagnostic is too technical/complex. However, "extending several types at once" is not quite accurate as the overall composition might simplify to a single type.

varungandhi-apple replied:
How about

> extending protocol compositions and typealiases of protocol compositions may lead to unexpected errors while type-checking. Instead, pick a single type to be extended and use a 'where' clause for additional constraints.

Is that better?

theblixguy replied:
Yeah, that’s looks much better to me. Perhaps the last bit could be “checking; choose a single type...”.

slavapestov replied:
I think "this may lead to unexpected errors while type-checking" sounds a bit ominous. I would prefer if the diagnostic just said "extending a protocol composition is not supported; extending '%0' instead" where %0 is the first nominal.

theblixguy commented on lib/AST/NameLookup.cpp:
```diff
@@ -2170,7 +2171,18 @@ ExtendedNominalRequest::evaluate(Evaluator &evaluator,
+    // If there were several references, we will report an ambiguity error
+    // anyways, so this diagnostic is not applicable.
+    if (referenced.size() == 1)
+      // TODO(Varun): How do we get the source range for the extended type?
```
Could you grab it from the extended type repr? (i.e. `ext->getExtendedTypeRepr().getSourceRange()`?)

varungandhi-apple replied:
IIUC, that's not guaranteed to succeed. The same function has a check earlier:
```
  // Prefer syntactic information when we have it.
  if (auto typeRepr = ext->getExtendedTypeRepr()) {
```
so it is possible that we do not have the syntactic information at this point in time.

theblixguy replied:
Is it the same with the extended nominal type? I suppose you can skip highlighting the location if it’s not available (unless it’s always not available at this stage).

slavapestov replied:
You might want to diagnose this in ExtendedTypeRequest::evaluate() instead of doing it here, because there's no reason to emit the warning for extensions in secondary files.

Also now that you're serializing the extended nominal directly, you can fix the ClangImporter's synthesis of ExtensionDecls to also set the extended nominal, and remove the code from ExtendedNominalRequest for handling the case of a null TypeRepr. It won't happen for parsed extensions.

varungandhi-apple replied:
> You might want to diagnose this in ExtendedTypeRequest::evaluate() instead of doing it here, because there's no reason to emit the warning for extensions in secondary files.

Do we have an invariant that ExtendedTypeRequest::evaluate() is not called for extensions in secondary files but ExtendedNominalRequest::evaluate() is?

varungandhi-apple replied:
> Also now that you're serializing the extended nominal directly, you can fix the ClangImporter's synthesis of ExtensionDecls to also set the extended nominal, and remove the code from ExtendedNominalRequest for handling the case of a null TypeRepr. It won't happen for parsed extensions.

There's an additional case here here in ParseDecl which can lead to a null TypeRepr:

https://github.com/apple/swift/blob/27ded648cf330d32f421c7aadc1ec132f292f556/lib/Parse/ParseDecl.cpp#L3851-L3855

Any suggestions on how we might get rid of it?

slavapestov replied:
> Do we have an invariant that ExtendedTypeRequest::evaluate() is not called for extensions in secondary files but ExtendedNominalRequest::evaluate() might be?

It's not an invariant, but we want to minimize requests in secondary files where possible.

> There's an additional case here here in ParseDecl which can lead to a null TypeRepr:

This is the invalid case with on type written (`extension { ... }`). In this case ExtendedNominalRequest should just return null.

slavapestov commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -3732,10 +3734,10 @@ class swift::DeclDeserializer {
-                              std::move(baseTy));
-    auto nominal = extension->getExtendedNominal();
+                              std::move(extendedType));
+    auto nominal = dyn_cast<NominalTypeDecl>(MF.getDecl(extendedNominalID));
```
You need to call ctx.evaluator.cacheOutput() to cache the result of ExtendedNominalRequest as well, so that the request doesn't get re-computed.

varungandhi-apple replied:
I don't quite understand -- in some cases, caching is automatic whereas not in other cases? Or does it always have to be done manually?

jrose-apple commented:
Can we add a warning for composition extensions that escape this checking but *don't* cause problems—the ones that canonicalize to a single protocol? There's no reason for people to be spelling things the composition way in that case.

varungandhi-apple commented:
> Can we add a warning for composition extensions that escape this checking but don't cause problems

I thought the check was supposed to catch precisely those cases, but my implementation right now is incorrect. If the composition does not normalize to a single type, we emit an error anyways, so I think we should not emit a warning on top of that error.

jrose-apple commented:
Oh, right, I forgot about that error. But yeah, your test case should have the warning, then.

slavapestov commented:
> I thought the check was supposed to catch precisely those cases, but my implementation right now is incorrect. If the composition does not normalize to a single type, we emit an error anyways, so I think we should not emit a warning on top of that error.

Since we don't know that the type canonicalizes down to a single term until we compute the extended type (not the extended nominal), it makes sense to move the warning, either to ExtendedTypeRequest, or DeclChecker::visitExtensionDecl().

varungandhi-apple committed:
[NFC] Replace tabs with spaces.

Does this make me a space engineer?

varungandhi-apple committed:
Serialize extended nominal separately when serializing an extension.

Instead of computing it from the extended type after deserialization --
which is tricky to do, due to potential presence of protocol
compositions -- we obtain the extended nominal directly.

Fixes SR-11227 and linked rdar://problem/53712389.

varungandhi-apple committed:
[NFC] Always use syntactic information in ExtendedNominalRequest.

Since we started saving the nominal in the .swiftmodule, changing the other
call-sites in ClangImporter to save the value means that we can rely on
the typeRepr being non-null except when the syntax tree was ill-formed.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
I've addressed the suggestions, should be good to review again.

slavapestov commented on lib/Serialization/Serialization.cpp:
```diff
@@ -2652,12 +2652,12 @@ class Serializer::DeclSerializer : public DeclVisitor<DeclSerializer> {
     // parameters.
-    auto baseNominal = baseTy->getAnyNominal();
-    (void)S.addDeclRef(baseNominal);
+    auto firstNominalInExtendedType = extendedType->getAnyNominal();
```
Why not just extension->getExtendedNominal()?

slavapestov reviewed (approved):
Looks good, only a few minor comments

varungandhi-apple commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -1601,6 +1601,10 @@ ERROR(inferred_opaque_type,none,
       "non-nominal type %0 cannot be extended", (Type))
+WARNING(composition_in_extended_type,none,
+        "extending a protocol composition is not supported; extending %0 "
+        "instead. Use a 'where' clause for additional constraints.",
```
We try to avoid multiple sentences in diagnostics (see docs/Diagnostics.md). Do we really need the `where` clause bit?

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3075,14 +3075,34 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+          || (extTypeNominal && extTypeNominal != nominal))
+        ED->diagnose(diag::composition_in_extended_type,
+                     nominal->getDeclaredType())
+          .highlight(extTypeRepr->getSourceRange());
```
Can we add a fix-it to use the type that the compiler picked?

varungandhi-apple replied:
Any suggestions for the fix-it message? How about "would you like to replace the composition with %0"?

theblixguy replied:
Maybe “replace composition type %0 with %1”?

jrose-apple replied:
No message, just a fix-it directly on the warning. It's always the right thing to do in the case where the warning comes up, right?

jrose-apple replied:
(Or rather, the "extending %0 instead" serves as the message.)

varungandhi-apple replied:
> It's always the right thing to do in the case where the warning comes up, right?

Nope. That's why I didn't have a fix-it earlier. It is unclear if the user wants to extend the more specific type or the less specific type. If I were to hazard a guess, I would guess that they want to extend the more specific type.

We are treating it as if they want to extend whatever came first (and not the most specific type), which is potentially not what they want, and can lead to a perplexing type error (as I mentioned in the earlier error message but I got rid of that).

varungandhi-apple replied:
If "right thing to do" == "same dynamic semantics", the yes. If "right thing to do" == "matches user expectation on what code should do", then potentially no.

jrose-apple replied:
:-/ Good point. Then yeah, it would be a separate note. "did you mean to extend the most-inherited protocol %0?" perhaps?

("most-derived" and "refining" aren't quite user-visible terms. Looking at [the Swift Programming Language book](https://docs.swift.org/swift-book/LanguageGuide/Protocols.html) "inherit" is the only term that's really used. "most-specific" might work too.)

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
@swift-ci please smoke test

jrose-apple commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3075,14 +3075,40 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
+        if (firstNominalIsNotMostSpecific)
+          ED->diagnose(diag::composition_in_extended_type_alternative,
+                       extTypeNominal->getDeclaredType());
+      }
```
Fix-its? On both branches of the `if`, I guess.

varungandhi-apple replied:
Shouldn't we _not_ be providing a fix-it since there isn't one clear right answer? :confused:

jrose-apple replied:
A fix-it on a note counts as an option, and it's definitely the most *likely* right answer in the case where you print the note. When you don't print the note (because the compiler picked the most specific type anyway), it's almost certainly the right answer.

varungandhi-apple commented:
@swift-ci please smoke test

varungandhi-apple commented:
> /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/cmake-linux-x86_64/bin/cmake --build /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/cmark-linux-x86_64 -- -j16 test
[0/1][  0%][0.000s] Running tests...
Test project /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-master/buildbot_linux/cmark-linux-x86_64
    [...]
    Start 4: pathological_tests_library
4/9 Test # 4: pathological_tests_library .......***Failed   12.66 sec

What on earth does my change have to do with cmark ☹️

jrose-apple commented:
I don't think that can possibly affect it. We also barely modify our fork of CMark, so it's unlikely the Swift project will be fixing whatever happened here. (I guess it's just as likely a hiccup in our Jenkins setup, too.)

If you rerun the Linux tests it will almost certainly vanish, unless a whole bunch of people have hit this since yesterday.y

varungandhi-apple commented:
@swift-ci please smoke test and merge

varungandhi-apple commented:
@swift-ci please smoke test Linux platform

varungandhi-apple commented:
Phew!
